ops/model.py

```diff
@@ -1348,6 +1348,51 @@ class InvalidStatusError(ModelError):
     """Raised if trying to set an Application or Unit status to something invalid."""
 
 
+def _create_action_set_dict(dictionary, parent_key=False):
+    """Turn a nested dictionary into a flattened dictionary, using '.' as a key seperator.
+
+    This is used to allow nested dictionaries to be translated into the dotted format required by
+    the Juju `action-set` hook tool in order to set nested data on an action.
+
+    Additionally, this method performs some validation on keys to ensure they only use permitted
+    characters, which according to the action-set docs, are lowercase characters, digits, hyphens,
+    and periods.
+
+    Example::
+
+        test_dict = {'a': {'b': 1, 'c': 2}}
+        flat_dict = flatten(test_dict)
+        # flat_dict is now {'a.b': 1, 'a.c': 2}
+
+    Arguments:
+        dictionary: The dictionary to flatten
+        parent_key: The string to prepend to dictionary's keys
+
+    Returns:
+        A flattened dictionary
+
+    Raises:
+        ValueError: if a dict is passes with a key that fails to meet the format requirements
+    """
+    items = []
+    for key, value in dictionary.items():
+        # Ensure the key is of a valid format, and raise a ValueError if not
+        if not re.match(r'^[0-9a-z](([A-Za-z0-9-]+)?[a-z0-9])?$', key):
```

Comment:
re.compile is [slightly](https://stackoverflow.com/a/13640709/3516684) faster

Reply:
Additionally, the regex needs a few tweaks: as sed-i mentioned there's no period, and I see that it allows uppercase letters in the middle, whereas action-get's docs on 2.9.14 indicate that lowercase is required.

I'd suggest this regex instead:

```python3
        if not re.match(r'^[a-z0-9](([a-z0-9-.]+)?[a-z0-9])?$', key):
```

I also agree with the re.compile feedback; it's a small change and has the potential to be faster, given the loop here.  Maybe not by much, granted.

Reply:
So I deliberately left the period out, as we'll be inserting periods as we split the dicts up - we're validating the keys of normal dicts here. I suppose we could allow periods in here too, but it could lead to additional, unexpected nesting?

As for uppercase, I interpreted the docs that the first and last letters must be lowercase/numeric, but uppercase should be allowed in the middle, as well as hyphens? @petevg may be able to confirm?

I'll change to `re.compile`, though. I agree that's an easy fix. I'll also bust the validation out into its own method, I think.